procps: enable systemd support when available

binarin opened the pull request:
###### Motivation for this change

`ps` can show various systemd-related information, like a unit to which a process belongs. But when it's not compiled in, it shows only `?` in such fields.

Can be tested with:

    ps -o unit= ax

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

matthewbauer commented on pkgs/os-specific/linux/procps-ng/default.nix:
```diff
@@ -1,4 +1,6 @@
-{ lib, stdenv, fetchurl, ncurses, pkgconfig }:
+{ lib, stdenv, fetchurl, ncurses, pkgconfig
+, systemd ? null
```
This is not the correct way to do this. For instance on macOS systems systemd will not be null - it just won't build because it is linux-only. The closest to a correct solution is something like this:

```
{ lib, stdenv, fetchurl, ncurses, pkgconfig
, withSystemd ? stdenv.isLinux, systemd
```

binarin commented:
Something strange happens with this change, it can be seen in GrahamcOfBorg error message above. And it actually breaks nixos VM tests, with the same message.

I've tried to look into this a bit, and it looks like things like `fetchzip` and `fetchpatch` are getting `fetchurlBoot` instead of full-fledged `fetchurl` (where support for all the fancy options really is).

binarin commented:
I don't know how to investigate that `fetchurl`/bootstrap issue. So for now I've made this option available only by explicit request, so that https://github.com/NixOS/nixpkgs/pull/34265 can move forward.

matthewbauer commented on pkgs/os-specific/linux/procps-ng/default.nix:
```diff
@@ -11,14 +22,16 @@ stdenv.mkDerivation rec {
   buildInputs = [ ncurses ];
-  nativeBuildInputs = [ pkgconfig ];
+  nativeBuildInputs = [ pkgconfig ]
+    ++ lib.optional withSystemd systemd;
```
This should be a build input.

matthewbauer commented:
I pushed a few changes. I could not reproduce the ofborg failure. It could have been something that was closed.

binarin commented:
I've dug deeper into this issue, as VM tests were still failing. Short story is that during tests `procps`/`systemd` is evaluated with stdenv `bootstrap-stage4-stdenv-linux` that is not functional enough to build `systemd`.

- Everything starts here https://github.com/NixOS/nixpkgs/blob/43650115e34ea610e53f718d9e02de7222ac0ba2/nixos/modules/misc/documentation.nix#L81-L82
- This pulls in texinfo with interactive flag set
- Looking at texinfo history, I can see that procps dependency (conditionalized on doCheck) was introduced by https://github.com/NixOS/nixpkgs/pull/11527 - https://github.com/NixOS/nixpkgs/blob/aa5184a88dd116b80aa9d75772db860a2dbd431a/pkgs/development/tools/misc/texinfo/5.2.nix#L15
- So the only reason for procps to be pulled in is to run tests, but right now it always happen
- When I start tests using
    ```
    NIX_PATH=nixpkgs=$(pwd) nix-build '<nixpkgs/nixos/tests/boot.nix>'
    ```
  all fails
- Here is the snippet of what chain of derivations is causing this: https://gist.github.com/binarin/9f4cb48211eecf9040cd9269326d32f1
  What's going on here is that `dbus` systemd service tries to pull in `system-path`, which now includes the `texinfoInteractive`
- And for some reason `procps` is evaluated with `bootstrap-stage4-stdenv-linux` stdenv, which is not functional enough to build systemd

As a quick hack I'm going to make a systemd dependency conditional not only on `isLinux`, but also on the absence of `bootstrap` in `stdenv.name`.
Or maybe somebody with more understanding of bootstrap stage layering, fix-points, etc. will chime in with a proper fix.

binarin committed:
procps: enable systemd support when available

`ps` can show various systemd-related information, like a unit to
which a process belongs. But when it's not compiled it, it shows only
'?' in such fields.

Can be tested with:

    ps -o unit= ax

vcunat commented:
Merged along with 5ae438a3e.
